components/modules/mqtt.c

```diff
@@ -164,572 +156,633 @@ static esp_err_t mqtt_event_handler(esp_mqtt_event_handle_t event)
-		int res = lua_pcall( L, 1, 0, 0 ); //call the connect callback
-		if( res != 0 )
-			NODE_DBG("CB:connect: Error when calling connect callback - (%d) %s\n", res, luaL_checkstring( L, -1 ) );
-	}
-	lua_settop(L, top);
-	event_free(event);
-}
+// task_connected is run when the mqtt client connected
+static void task_connected(task_param_t param, task_prio_t prio) {
+    lua_State* L = lua_getstate();  //returns main Lua state
+    if (L == NULL)
+        return;
+
+    // extract the event data out of the task param
+    esp_mqtt_event_handle_t event = (esp_mqtt_event_handle_t)param;
+
+    // recover the mqtt context from the event user_context field:
+    mqtt_context_t* mqtt_context = *(mqtt_context_t**)event->user_context;
+    NODE_DBG("CB:connect: state %p, settings %p, stack top %d\n", L, event->client, lua_gettop(L));
+
+    event_free(event);  // free the event copy memory
 
-static void _disconnected_cb(task_param_t param, task_prio_t prio)
-{
-	lua_State * L = lua_getstate(); //returns main Lua state
-	if( L == NULL )
-		return;
-
-	esp_mqtt_event_handle_t event = (esp_mqtt_event_handle_t) param;
-
-	int top = lua_gettop(L);
-	lua_checkstack(L, 8);
-
-	char key[64];
-	snprintf(key, 64, "mqtt_%p", event->client);
-	lua_getglobal( L, key ); //retrieve MQTT table from _G
-	NODE_DBG("CB:disconnect: state %p, settings %p, stack top %d\n", L, event->client, lua_gettop(L));
-
-	lua_getfield( L, -1, "_connect_nok" );
-	if( lua_isfunction( L, -1 ) )
-	{
-		NODE_DBG("CB:disconnect: calling registered one-shot disconnect callback\n");
-		lua_pushvalue( L, -2 ); //dup mqtt table
-		int res = lua_pcall( L, 1, 0, 0 ); //call the disconnect callback
-		if( res != 0 )
-			NODE_DBG("CB:disconnect: Error when calling one-shot disconnect callback - (%d) %s\n", res, luaL_checkstring( L, -1 ) );
-
-		//after connecting ok, we clear _both_ the one-shot callbacks
-		lua_pushnil(L);
-		lua_setfield(L, 1, "_connect_ok");
-		lua_pushnil(L);
-		lua_setfield(L, 1, "_connect_nok");
-	}
-
-	// now we check for the standard connect callback registered with 'mqtt:on()'
-	lua_getfield( L, -1, "_on_offline" );
-	if( !lua_isfunction( L, -1 ) || lua_isnil( L, -1 ) )
-	{
-		event_free(event);
-		return;
-	}
-
-	NODE_DBG("CB:disconnect: calling registered standard offline callback\n");
-	lua_pushvalue( L, -2 ); //dup mqtt table
-	int res = lua_pcall( L, 1, 0, 0 ); //call the offline callback
-	if( res != 0 )
-		NODE_DBG("CB:disconnect: Error when calling offline callback - (%d) %s\n", res, luaL_checkstring( L, -1 ) );
-
-	lua_settop(L, top);
-	event_free(event);
+    if (mqtt_context->self <= 0) {  // if this reference is unset something weird is happening
+        NODE_DBG("CB:connect: Received event on a collected object\n");
+        return;
+    }
+
+    int top = lua_gettop(L);  // save the stack status to restore it later
+    lua_checkstack(L, 3);     // make sure there are at least 3 slots available
+
+    // pin our object by putting a reference on the stack,
```

Comment:
Indeed, without this our average Joe would break our implementation this way:

```lua
client = mqtt.Client("test", 100)
client:connect(
    "1.2.3.4",
    1883,
    0,
    function()
        print("connected")
    end,
    function()
        print("failed")
        client = nil -- this can cause client to be garbage collected *before*
                     -- the rest of the event handler back in C has had the chance to execute, 
                     -- wreaking havoc, and of course rebooting your device or something worse.
    end
)
```